man/man1/cvrtclm.1
.TH cvrtclm 1  "USER COMMANDS"
.SH NAME
cvrtclm \- convert version of clm file to version 3 or 4
.SH SYNOPSIS
.B cvrtclm
[\-4] [\-type {byte|short|int|float|double}] [\-scale s] [\-cellsize size] [\-swapnstep] [-timestep n]
.I infile outfile
.SH DESCRIPTION
Program converts clm file into clm file with version 3 or 4. For version 1 file cellsize and scaling factor have to be specified, if they differ from default values. For version 2 type has to be specified if different from short.
.SH OPTIONS
.TP
\-4
Convert to version 4 clm file, default is 3.
.TP
\-type {byte|short|int|float|double}
Set type of data in header, default is short.
.TP
\-scale s
Set scaling factor in header, default is 1.0.
.TP
\-cellsize size
Set cell size in header, default is 0.5.
.TP
\-swapnstep
Swap nbands for nstep for version 4 clm files.
.TP
\-timestep n
Set time step for version 4 clm files, default is 1.
.TP
.I infile
Filename of clm file to convert.
.TP
.I outfile
Filename of clm data file written. Must differ from infile.
.SH EXAMPLES
.TP
Convert clm wateruse data file to version 3:
.B cvrtclm
-type int wateruse_1900-2000.bin wateruse_1900-2000_v3.clm
.PP
.SH EXIT STATUS
.B cvrtclm
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), asc2clm(1), cru2clm(1), grid2clm(1), cft2clm(1), catclm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code reviewe: Stephen Wirth (wirth@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de)

### Added

- Utility `restart2yaml` added to convert the new restart files into human readable YAML or JSON files.
- Utility `json2restart` added to convert JSON files into restart files.
- alias `restart2json` added for `restart2yaml -json`.
- Output of time spent in reading and writing the restart file added.

### Changed

- Restart file format changed to include metadata. A binary JSON-like structure has been implemented. The order of the objects read can be different from the order objects have been written, but performance of reading can be degraded.
- Utility `lpjcat` updated to support the new restart file format.


## [5.9.25] - 2025-05-22

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson` instead of being hard coded.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.
- Output format for cell size updated in `fprintheader.c`, `fprintoutputjson.c`, `fprintjson.c`, and `openinputfile.c`.

### Added

- New utility `splitclm` added to copy specific bands from a clm file.
- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `cdf2clm`, `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Options `-zero` and `-search` added to `regridlpj`.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`, `txt2clm` and `bin2cdf`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.
- Check for correct setting of nstep and nbands added in `txt2clm`.

### Fixed

- Misspelled `-fsanitize=address` option corrected in `Makefile.mpich`.
- Missing `free(cmd)` added in `fprintincludes.c`.
- Parsing of options corrected in `fprintincludes.c`.
- Utility `bin2cdf` fixed for timestep>1.


## [5.9.24] - 2025-05-15

### Contributors

- author: David Hötten (davidho@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Added

- Merge request template ``.gitlab/merge_request_templates/default.md`` that includes several checklists.


## [5.9.23] - 2025-05-08

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Fabian Stenzel(stenzel@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and undefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in utilities.
- Prescribed landcover input file changed in `input.cjson` to a file with the same number of natural PFTs.
- All remaining unsafe calls to `sprintf()` replaced by calls to `snprinf()` or `getsprinf()`.

### Fixed

- Check for missing values for integer input corrected in `readclimate_netcdf.c`.
- Memory leak fixed in `filesexist.c`.
- Calculation of global area fixed in `printglobal.c`.
- Check for null attribute pointer added in `fprintjson.c`.
- Uninitialized grid name initialized in `mathclm.c`.
- Call to `snprintf()` replaced by `getsprintf()` in `reservoir2cdf.c`.
- Datatype for index and writing the header corrected in `lpjcat.c`.
- Functions for qsort(), bisect(), and leftmostzero() changed to avoid run-time errors with `-check` option.
- Pointer set to NULL  if `k_est`is NULL in `initmanage.c`.
- Missing argument added in opening aquifer file for IMAGE.
- Scaling of coordinates fixed in `joingrid.c`.
- Fraction pointer initialized to NULL in `freadresdata.c` to avoid SEGV in `lpjprint.c` if reservoir data is read from restart file.
- Index set correctly for outflow cells in `regriddrain.c`.
- Check for maximum discharge length corrected in `printdrain.c`.
- Number of years of clm file set to 1 in `grid2clm.c`.
- Function `fscanlandcovermap.c` corrected (issue https://github.com/PIK-LPJmL/LPJmL/issues/48).
- Access out of bounds for `soil->freezdepth` fixed in `soiltemp.c`.
- Format specifier changed in `cat2bsq.c` and `cutclm.c` to compile without warnings using the clang compiler (issue #372).


## [5.9.22] - 2025-03-14

### Contributors


src/utils/cvrtclm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                       c  v  r  t  c  l  m  .  c                                \n**/
/**                                                                                \n**/
/**     Program converts CLM file to latest version 3                              \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-4] [-scale s] [-type {byte|short|int|float|double}] [-timestep n] [-cellsize size] [-swapnstep] infile outfile\n"

#define BUFSIZE (1024*1024) /* size of read buffer */

int main(int argc,char **argv)
{
  FILE *infile,*outfile;
  Header header;
  int version,new_version,timestep;
  String id;
  const char *progname;
  char *endptr;
  int iarg,*index;
  size_t i,size;
  long long filesize;
  void *buffer;
  Type datatype;
  float scalar,cellsize;
  Bool swap,swapnstep;
  /* set default values */
  datatype=LPJ_SHORT;
  scalar=1;
  cellsize=0.5;
  new_version=3;
  swapnstep=FALSE;
  timestep=1;
  progname=strippath(argv[0]);
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-scale"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Argument missing for option '-scale'.\n");
          return EXIT_FAILURE;
        }
        scalar=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-scale'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(scalar<=0)
        {
          fputs("Scale factor less than or equal to zero.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-4"))
        new_version=4;
      else if(!strcmp(argv[iarg],"-swapnstep"))
        swapnstep=TRUE;
      else if(!strcmp(argv[iarg],"-type"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Argument missing for option '-type'.\n");
          return EXIT_FAILURE;
        }
        i=findstr(argv[++iarg],typenames,N_TYPES);
        if(i==NOT_FOUND)
        {
          fprintf(stderr,"Invalid argument '%s' for option '-type'.\n"
                  USAGE,argv[iarg],progname);
          return EXIT_FAILURE;
        }
        datatype=(Type)i;
      }
      else if(!strcmp(argv[iarg],"-cellsize"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Argument missing for option '-cellsize'.\n");
          return EXIT_FAILURE;
        }
        cellsize=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-cellsize'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(cellsize<=0)
        {
          fputs("Cell size less than or equal to zero.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-timestep"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Argument missing for option '-timestep'.\n");
          return EXIT_FAILURE;
        }
        timestep=(float)strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-timestep'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(timestep<=0)
        {
          fputs("Time step less than or equal to zero.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[iarg],progname);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+2)
  {
    fprintf(stderr,"Argument(s) missing.\n"
            USAGE,progname);
    return EXIT_FAILURE;
  }
  if(!strcmp(argv[iarg],argv[iarg+1]))
  {
    fputs("Error: source and destination filename are the same.\n",stderr);
    return EXIT_FAILURE;
  }
  infile=fopen(argv[iarg],"rb");
  if(infile==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s.\n",argv[iarg],strerror(errno));
    return EXIT_FAILURE;
  }
  version=READ_VERSION;
  if(freadanyheader(infile,&header,&swap,id,&version,TRUE))
  {
    fprintf(stderr,"Error reading header in '%s'.\n",argv[iarg]);
    return EXIT_FAILURE;
  }
  if(version==1)
  {
    header.scalar=scalar;
    header.cellsize_lon=header.cellsize_lat=cellsize;
  }
  if(version<3)
    header.datatype=datatype;
  if(new_version==4 && swapnstep)
  {
     header.nstep=header.nbands;
     header.nbands=1;
  }
  header.timestep=timestep;
  filesize=getfilesizep(infile);
  if((header.order==CELLINDEX && filesize!=sizeof(int)*header.ncell+(long long)header.ncell*header.nbands*header.nstep*header.nyear*typesizes[header.datatype]+headersize(id,version)) ||
     (header.order!=CELLINDEX && filesize!=(long long)header.ncell*header.nbands*header.nstep*header.nyear*typesizes[header.datatype]+headersize(id,version)))
  {
     fprintf(stderr,"Error: File size of '%s' does not match nyear*nbands*nstep*ncell.\n",argv[iarg]);
     fclose(infile);
     return EXIT_FAILURE;
  }
  outfile=fopen(argv[iarg+1],"wb");
  if(outfile==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[iarg+1],strerror(errno));
    return EXIT_FAILURE;
  }
  fwriteheader(outfile,&header,id,new_version);
  buffer=malloc(BUFSIZE);
  if(buffer==NULL)
  {
    printallocerr("buffer");
    return EXIT_FAILURE;
  }
  size=filesize-headersize(id,version);
  if(header.order==CELLINDEX)
  {
    index=newvec(int,header.ncell);
    if(index==NULL)
    {
      printallocerr("index");
      free(buffer);
      return EXIT_FAILURE;
    }
    size-=sizeof(int)*header.ncell;
    freadint(index,header.ncell,swap,infile);
    if(fwrite(index,sizeof(int),header.ncell,outfile)!=header.ncell)
    {
      free(index);
      free(buffer);
      fprintf(stderr,"Error writing data in '%s'.\n",argv[iarg+1]);
      return EXIT_FAILURE;
    }
    free(index);
  }
  for(i=0;i<size / BUFSIZE;i++)
  {
    switch(typesizes[header.datatype])
    {
      case 2:
        freadshort(buffer,BUFSIZE/2,swap,infile);
        break;
      case 4:
        freadint(buffer,BUFSIZE/4,swap,infile);
        break;
      case 8:
        freadlong(buffer,BUFSIZE/8,swap,infile);
        break;
      default:
        fread(buffer,1,BUFSIZE,infile);
    }
    if(fwrite(buffer,1,BUFSIZE,outfile)!=BUFSIZE)
    {
      free(buffer);
      fprintf(stderr,"Error writing data in '%s'.\n",argv[iarg+1]);
      return EXIT_FAILURE;
    }
  }
  if(size % BUFSIZE>0)
  {
    switch(typesizes[header.datatype])
    {
      case 2:
        freadshort(buffer,(size % BUFSIZE)/2,swap,infile);
        break;
      case 4:
        freadint(buffer,(size % BUFSIZE)/4,swap,infile);
        break;
      case 8:
        freadlong(buffer,(size % BUFSIZE)/8,swap,infile);
        break;
      default:
        fread(buffer,1,size % BUFSIZE,infile);
    }
    if(fwrite(buffer,1,size % BUFSIZE,outfile)!=size % BUFSIZE)
    {
      free(buffer);
      fprintf(stderr,"Error writing data in '%s'.\n",argv[iarg+1]);
      return EXIT_FAILURE;
    }
  }
  free(buffer);
  fclose(infile);
  fclose(outfile);
  return EXIT_SUCCESS;
} /* of 'main' */
